src/native/libs/System.Security.Cryptography.Native/pal_evp_pkey.c

```diff
@@ -89,7 +89,220 @@ int32_t CryptoNative_UpRefEvpPkey(EVP_PKEY* pkey)
+    // recomputing them in scratch integers.
+    BIGNUM* p1 = NULL;
+    BIGNUM* q1 = NULL;
+    int ret = 0;
+
+    RSA_get0_key(rsa, &n, &e, &d);
+
+    // Always need public parameters.
+    if (!n || !e)
+    {
+        ERR_PUT_error(ERR_LIB_RSA, 0, RSA_R_VALUE_MISSING, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // Compatibility: We put this error for OpenSSL 1.0.2 and 1.1.x when the modulus is zero because OpenSSL did not
+    // handle this correctly. Continue to use the same error if the modulus is zero.
+    if (BN_is_zero(n))
+    {
+        ERR_put_error(ERR_LIB_EVP, 0, EVP_R_DECODE_ERROR, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // OpenSSL has kept this value at 16,384 for all versions.
+    if (BN_num_bits(n) > OPENSSL_RSA_MAX_MODULUS_BITS)
+    {
+        ERR_PUT_error(ERR_LIB_RSA, 0, RSA_R_MODULUS_TOO_LARGE, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // Exponent cannot be 1 and must be odd
+    if (BN_is_one(e) || !BN_is_odd(e))
+    {
+        ERR_PUT_error(ERR_LIB_RSA, 0, RSA_R_BAD_E_VALUE, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // At this point everything that is public has been checked. Mark as successful and clean up.
+    if (isPublic)
+    {
+        ret = 1;
+        goto done;
+    }
+
+    // We do not support validating multi-prime RSA. If there are extra primes (more than two) then treat it as a
+    // decoding failure.
+    if (RSA_get_multi_prime_extra_count(rsa) != 0)
+    {
+        ERR_put_error(ERR_LIB_EVP, 0, EVP_R_DECODE_ERROR, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // Get the private components now that we've moved on to checking the private parameters.
+    RSA_get0_factors(rsa, &p, &q);
+
+    // Need all the private parameters now.
+    if (!d || !p || !q)
+    {
+        ERR_PUT_error(ERR_LIB_RSA, 0, RSA_R_VALUE_MISSING, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // Set up the scratch integers
+    if ((ctx = BN_CTX_new()) == NULL ||
+        (x = BN_new()) == NULL ||
+        (y = BN_new()) == NULL ||
+        (p1 = BN_new()) == NULL ||
+        (q1 = BN_new()) == NULL)
+    {
+        goto done;
+    }
+
+    // multiply p and q and put the result in x.
+    if (!BN_mul(x, p, q, ctx))
+    {
+        goto done;
+    }
+
+    // p * q == n
+    if (BN_cmp(x, n) != 0)
+    {
+        ERR_PUT_error(ERR_LIB_RSA, 0, RSA_R_N_DOES_NOT_EQUAL_P_Q, __FILE__, __LINE__);
+        goto done;
+    }
+
+    // Checking congruence of private parameters.
+    // de = 1 % lambda(n)
+    // lambda(n) = lcm(p-1, q-1)
+    // lambda(n) is known to be lambda(pq) already.
+    // p1 = p-1
+    // q1 = q-1
+    // x = lcm(x, y)
+    if (!BN_sub(p1, p, BN_value_one()) || !BN_sub(q1, q, BN_value_one()) || !Lcm(p1, q1, ctx, x))
+    {
+        goto done;
+    }
+
+    // de % lambda(n)
+    // put the result in y
+    if (!BN_mod_mul(y, d, e, x, ctx))
+    {
+        goto done;
+    }
+
```

Comment:
I only know of two computations for `d` (assuming the post-1979 interpretation of choosing a fixed `e` and calculating `d`), d-phi and d-lambda.  Phi(n) is less than N, and Lambda(p-1, q-1) is less than N because N == p*q.  So, in all cases, we know that D < N.

Is that worth adding a consistency check over?  Or, if someone built an RSAPrivateKey value that made d-big (d-lambda + 20*lcm(p-1, q-1)) are we OK with that?  (We're not from an RSAParameters perspective, but might be at a key-import, because, well, I don't remember which way we end up at the final call into OpenSSL)

Reply:
I'm not sure I care too much, but it's a single call to `BN_cmp` so, not difficult to add.